numaprom/watcher.py

```diff
@@ -0,0 +1,137 @@
+import os
+import time
+from functools import cache
+from typing import Optional
+
+from omegaconf import OmegaConf
+from watchdog.observers import Observer
+from numalogic.config import NumalogicConf
+from watchdog.events import FileSystemEventHandler
+
+from numaprom._constants import CONFIG_DIR, DEFAULT_CONFIG_DIR
+from numaprom import NumapromConf, get_logger, AppConf, MetricConf, UnifiedConf
+
+_LOGGER = get_logger(__name__)
+
+
+class ConfigManager:
+    config = {}
+
+    @staticmethod
+    def load_configs():
+        schema: NumapromConf = OmegaConf.structured(NumapromConf)
+
+        conf = OmegaConf.load(os.path.join(CONFIG_DIR, "config.yaml"))
+        app_configs = OmegaConf.merge(schema, conf).configs
+
+        conf = OmegaConf.load(os.path.join(DEFAULT_CONFIG_DIR, "config.yaml"))
+        default_configs = OmegaConf.merge(schema, conf).configs
+
+        conf = OmegaConf.load(os.path.join(DEFAULT_CONFIG_DIR, "numalogic_config.yaml"))
+        schema: NumalogicConf = OmegaConf.structured(NumalogicConf)
+        default_numalogic = OmegaConf.merge(schema, conf)
+
+        return app_configs, default_configs, default_numalogic
+
+    def update_configs(self):
+        app_configs, default_configs, default_numalogic = self.load_configs()
+
+        self.config["app_configs"] = dict()
+        for _config in app_configs:
+            self.config["app_configs"][_config.namespace] = _config
+
+        self.config["default_configs"] = dict(map(lambda c: (c.namespace, c), default_configs))
+        self.config["default_numalogic"] = default_numalogic
+
+        _LOGGER.info("Successfully updated configs - %s", self.config)
+        return self.config
+
+    @cache
+    def get_app_config(self, metric: str, namespace: str) -> Optional[AppConf]:
```

Comment:
Good question, this gets invalidated as soon as its called with different variables. So, this is only useful when we are making the same call multiple times in the same instance.

Reply:
I think we don't need an expiry, since it would be the same without any new changes to config. I just added a line to handler, to clear the cache whenever there is a change in config.